### PR TITLE
Add retries on 133 errors

### PR DESCRIPTION
**Pull request by bbergshaven**
I've seen a lot of 133 errors where the connectionState gets set to STATE_DISCONNECTED before the discovery is complete. By adding your reconnection logic to this case as well I've been able to almost remove the problem.

I've also increased the number of connection attempts from 2 to 3. This should actually be a configurable number, so the Intent EXTRA_ATTEMPT could be changes to EXTRA_NUM_ATTEMPTS and the service could decrement this number. But this works for now and makes the DFU library more stable.

### CODE REVIEW

File: dfu/src/main/java/no/nordicsemi/android/dfu/DfuBaseService.java [hunk 1]
```diff
@@ -1007,7 +1031,9 @@ protected void onHandleIntent(final Intent intent) {
 					logi("Restarting the service");
 					final Intent newIntent = new Intent();
 					newIntent.fillIn(intent, Intent.FILL_IN_COMPONENT | Intent.FILL_IN_PACKAGE);
-					newIntent.putExtra(EXTRA_ATTEMPT, 1);
+                    int attempts = intent.getIntExtra(EXTRA_ATTEMPT,0);
```
Comment (nachtien) [inline dfu/src/main/java/no/nordicsemi/android/dfu/DfuBaseService.java]: Make this code look pretty and indent it :)

**Comment by nachtien** [comment]
I'm seeing a ton of 133 errors on a Samsung Galaxy S6.  This occurred after we switched to DFU mode then attempted to reconnect to the peripheral.  Unfortunately, your PR did not fix this issue.  The way I ended up fixing this was in the Gatt callback, checking for the 133, and handling it there (in our case, just returning and not doing anything).

**Comment by philips77** [comment]
Hi,
The reason the service retries twice is to avoid Invalid State error thrown when DFU was restarted after it was stopped initially. It does not solve 133, however if error occurred repeating may work. Actually it's usually the only solution, except restarting Bluetooth or the phone.
But in most cases the 133 can be avoided by handling the connection parameters in some magic way. It is thrown, in my opinion, when some timing issues occur. We had once solved it by sending connection parameters update after connecting from the peripheral (don't know how and when).

**Comment by bbergshaven** [comment]
Hello,
I just discovered one thing that had a major impact on the amount of 133 errors we got while doing DFU. As you mention over the problem might be correlated with connection parameters and many before has mentioned that 133's can origin from timing issues.

What we saw was that after sending the necessary commands to put the device in bootloader modus and disconnecting, the act of connecting to the "new" (changed mac-address) device would randomly fail. After a full day of BLE sniffing and debugging I realised that when the connection failed, with an 133 error, Android had not sent any connection request over the air, so the whole "could not connect" or "got disconnected before service discovery" was generated internally in Android.

The solution that I tried was to do a short scan before trying to connect. I didn't care about the scan results, this was just a way to put the Bluetooth radio in another state. And it seems to work.

So the procedure is as follows:
1. Write to the buttonless dfu service and make the device reboot into boot loader modus
2. After getting the disconnected callback in android, do a short (500-1000ms) scan. 
3. Start the Nordic DFU service with the new mac-address. 
4. Wait for DFU to successfully complete.

This might help some others struggling with the same error. 
Please let me know if anyone has more info regarding this issue.

**Comment by philips77** [comment]
Please,, have a look at #170.

**Pull request closed without merging**